### ISSUE TITLE
[BUG]: `nb::c_contig` constraint is not enforced for non-contiguous PyTorch arrays

### ISSUE THREAD

**Issue by bathal1** [wjakob/nanobind#278]
### Problem description

When binding a function processing tensors, one can specify contiguity flags, like `nb::c_contig`. Nanobind is expected to ensure the input tensors are contiguous in memory before calling the function. 

However, when passing a non-contiguous PyTorch tensor to such a function will raise an error, instead of making it contiguous under the hood.

Modifying the [Nanobind example](https://github.com/wjakob/nanobind_example) project with the following function definition allows to reproduce this issue.

### Function binding
```c++
m.def("add", [](nb::ndarray<float, nb::c_contig> a, nb::ndarray<float, nb::c_contig> b) { return 0; }, "a"_a, "b"_a);
```

### Reproducer
```python
from nanobind_example import add
import drjit as dr
import torch
a = torch.ones((3,3))
b = dr.ones(dr.llvm.TensorXf, (3,3))

# DrJit tensors, works as expected
print(b,b)
print(b,b[::-1])

# PyTorch tensors, fails
print(add(a,a))  # Contiguous tensors, returns 0
print(add(a,a.T)) # a.T is non-contiguous, this fails
```

This raises the following issue
```
Traceback (most recent call last):
  File "<stdin>", line 11, in <module>
TypeError: add(): incompatible function arguments. The following argument types are supported:
    1. add(a: ndarray[dtype=float32, order='C'], b: ndarray[dtype=float32, order='C']) -> int

Invoked with types: torch.Tensor, torch.Tensor
```

**Comment by wjakob** [comment]
Just double-checking: wasn't the error in the cholespy repository about a crash? (Rather than a `TypeError` mentioned here).

**Comment by bathal1** [comment]
The bug mentioned a crash, but the reproducer produced this error, without crashing.

**Comment by wjakob** [comment]
Does that mean that the code in the original issue didn't have the annotation? Otherwise it sounds like there are two separate issues.

**Comment by bathal1** [comment]
The bindings in `cholespy` already had the `nb::c_contig` annotations, prior to this [issue](https://github.com/rgl-epfl/large-steps-pytorch/issues/15).

**Comment by wjakob** [comment]
Fixed in https://github.com/wjakob/nanobind/commit/23ea3202294a5f3758b6d496404a875527386555.

**Comment by wjakob** [comment]
@bathal1 : I released nanobind v1.5.2 -- you might want to re-release cholespy with those fixes.
